### PR TITLE
normalize mean aggregator output

### PR DESCRIPTION
**Pull request by danielmend**
goals:
- create a nice interface for normalization + logit scale for each aggregator without needing to hardcode it into each aggregator class

### CODE REVIEW

**Comment by iejMac** [comment]
let's generalize this a bit more - not just the mean aggregator but all aggregators should have an easy way to normalize and use logit scale through a common interface or wrapper

File: src/aggregation/aggregator_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    """
+    Class to wrap aggregators so we can control their normalization and logit scale
+    """
+    def __init__(self, aggregator, logit_scale=100.0, normalize=True):
+        super().__init__()
+        self.aggregator = aggregator
+        self.logit_scale = logit_scale
+        self.normalize = normalize
+
+    def forward(self, x):
```
Comment (iejMac) [inline src/aggregation/aggregator_wrapper.py:15]: instead, add boolean parameters for ```prenorm``` and ```postnorm``` . If prenorm true than normalize x before the aggregator, if postnorm true normalize after the aggregator (both can be true)

File: src/aggregation/aggregator_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        x = self.aggregator(x)
+        return self.logit_scale * F.normalize(
```
Comment (iejMac) [inline src/aggregation/aggregator_wrapper.py:17]: do it like here https://github.com/mlfoundations/open_clip/blob/a6231f8cde726797330f149f1ffcfa0e63a55872/src/open_clip/model.py#L201

Comment (iejMac) [reply src/aggregation/aggregator_wrapper.py:17]: as in return self.logit_scale.exp() and don't multiply it

File: src/aggregation/mean.py [hunk 3]
```diff
@@ -5,9 +5,9 @@
 """
 
 from torch import nn
-
+import torch.nn.functional as F
```
Comment (iejMac) [inline src/aggregation/mean.py:8]: remove since normalization will happen in VideoCLIP class

**Comment by iejMac** [comment]
to complete this PR you'll need to:

- [ ] address current reviews
- [ ] integrate this in factory.py so it returns a VideoCLIP object with the correct aggregator inside
- [ ] confirm you can save and load it easily
- [ ] confirm resuming works for training (dummy runs are fine with the validation data you already have)
- [ ] confirm all aggregators work in VideoCLIP class as expected

File: .gitignore [hunk 4]
```diff
@@ -127,3 +127,7 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+src/Untitled.ipynb
```
Comment (iejMac) [inline .gitignore:130]: Remove these

File: src/aggregation/factory.py [hunk 5]
```diff
@@ -53,5 +52,5 @@ def create_model(cfg_path, pretrained=''):
         else:
             logging.warning(f'Pretrained weights ({pretrained}) not found for model {model_name}.')
             raise RuntimeError(f'Pretrained weights ({pretrained}) not found for model {model_name}.')
-
-    return model, get_model_string(model_dict)
+
```
Comment (iejMac) [inline src/aggregation/factory.py:55]: remove these tabs

File: src/training/train.py [hunk 6]
```diff
@@ -26,20 +26,21 @@ def train_one_epoch(model_video, model_text, logit_scale, data, epoch, optimizer
 
     running_loss = 0.0
     for  i, batch in enumerate(dataloader):
+        
         step = num_batches_per_epoch * epoch + i
         scheduler(step)
-
+
```
Comment (iejMac) [inline src/training/train.py:32]: tabs

File: src/training/train.py [hunk 7]
```diff
@@ -90,12 +91,12 @@ def evaluate(model_video, model_text, logit_scale, data, epoch, args, tb_writer=
     all_video_features, all_text_features = [], []
     with torch.no_grad():
         for i, batch in enumerate(dataloader):
-
+
```
Comment (iejMac) [inline src/training/train.py:94]: tabs

File: src/training/train.py [hunk 8]
```diff
@@ -66,9 +67,9 @@ def train_one_epoch(model_video, model_text, logit_scale, data, epoch, optimizer
                     wandb.log({name: val, 'step': step}, step=step)
 
             running_loss = 0.0
+        break
```
Comment (iejMac) [inline src/training/train.py:70]: remove break

File: src/aggregation/factory.py [hunk 9]
```diff
@@ -32,7 +31,7 @@ def load_checkpoint(model, checkpoint_path, strict=True):
     return incompatible_keys
 
 
-def create_model(cfg_path, pretrained=''):
+def create_model(cfg_path, clip_model, pretrained=''):
```
Comment (iejMac) [inline src/aggregation/factory.py:34]: instead of passing in the clip_model, hard code it to what we had in main.py and remove initializing the CLIP model in main.py

this is closer to what we're going to have in the end

File: src/training/data.py [hunk 10]
```diff
@@ -326,8 +326,8 @@ def get_wds_dataset(args, emb_transform, is_train, epoch=0, floor=False):
         dataset,
         batch_size=None,
         shuffle=False,
-        num_workers=args.workers,
-        persistent_workers=True,
+        num_workers=0,#args.workers,
```
Comment (iejMac) [inline src/training/data.py:329]: remember put these back to original once you're done testing

File: src/aggregation/aggregator_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            text_embeddings = self.model_text(x).float()
+        return F.normalize(text_embeddings, dim=-1) if postnorm else text_embeddings
+
+    def encode_video(self, x, prenorm=True, postnorm=True):
+        x = F.normalize(x.float(), dim=-1) if prenorm else x
+        x = self.aggregator(x)
+        x = F.normalize(x.float(), dim=-1) if postnorm else x
+        return x, self.logit_scale.exp()
+
+    def forward(self, toks, video_embeddings, prenorm=True, postnorm=True):
```
Comment (iejMac) [inline src/aggregation/aggregator_wrapper.py:51]: switch order, video_embeddings first then text tokens

File: src/training/train.py [hunk 12]
```diff
@@ -35,19 +34,16 @@ def train_one_epoch(model_video, model_text, logit_scale, data, epoch, optimizer
...
+        loss = loss_func(video_embeddings, text_embeddings, logit_scale)
         running_loss += loss.item() # maybe this doesn't make sense
 
         loss.backward()
         nn.utils.clip_grad_norm_(model_video.parameters(), args.grad_clip) # clip grads
         optimizer.step()
 
         batch_count = i + 1
-        if is_master(args) and (batch_count % 100 == 0 or batch == num_batches_per_epoch):
+        if is_master(args) and (batch_count % 100 == 1 or batch == num_batches_per_epoch):
```
Comment (iejMac) [inline src/training/train.py:46]: remember to put back to original

File: src/aggregation/aggregator_wrapper.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        text_embeddings = self.encode_text(toks, postnorm=postnorm)
+        video_embeddings, logit_scale = self.encode_video(video_embeddings, prenorm=prenorm, postnorm=postnorm)
+        return text_embeddings, video_embeddings, logit_scale
```
Comment (iejMac) [inline src/aggregation/aggregator_wrapper.py:54]: also swap order we output them

File: .gitignore [hunk 14]
```diff
@@ -126,4 +126,4 @@ venv.bak/
 dmypy.json
 
 # Pyre type checker
-.pyre/
+.pyre/
```
Comment (iejMac) [inline .gitignore:129]: what is this?

File: src/training/main.py [hunk 15]
```diff
@@ -120,13 +95,13 @@ def main():
...
     random_seed(args.seed)
-    model_video, model_str = create_model(args.model)
-    model_video.to(args.device)
 
     clip_model, _, preprocess = open_clip.create_model_and_transforms("ViT-H-14", pretrained="laion2b_s32b_b79k")
-    model_text = CLIPTxt(clip_model)
-    model_text.attn_mask = model_text.attn_mask.to(args.device)
-    model_text = model_text.to(args.device)
+    
+
```
Comment (iejMac) [inline src/training/main.py:101]: remove one \n while you're at it

File: .gitignore (hunk 14 above)
Comment (danielmend) [reply .gitignore:129]: not sure, this was there before

File: src/training/train.py [hunk 16]
```diff
@@ -35,11 +34,8 @@ def train_one_epoch(model_video, model_text, logit_scale, data, epoch, optimizer
 
         optimizer.zero_grad()
 
-        video_embeddings = model_video(embeddings)
-        with torch.no_grad():
-            text_embeddings = model_text(toks).float()
-
-        loss = loss_func(video_embeddings, text_embeddings, logit_scale.exp())
+        video_embeddings, text_embeddings, logit_scale  = model_video(embeddings, toks, prenorm=False, postnorm=False)
```
Comment (iejMac) [inline src/training/train.py:37]: I think we do want both prenorm and postnorm for training.

1. We want the model to learn the normalized CLIP embedding distribution
2. Why not normalize output if that's the function we want to approximate in the first place

File: src/training/train.py [hunk 17]
```diff
@@ -90,13 +85,11 @@ def evaluate(model_video, model_text, logit_scale, data, epoch, args, tb_writer=
...
     with torch.no_grad():
         for i, batch in enumerate(dataloader):
-
             embeddings, toks = batch
             embeddings = embeddings.to(device, non_blocking=True)
             toks = toks.to(device, non_blocking=True)
 
-            video_embeddings = model_video(embeddings)
-            text_embeddings = model_text(toks).float()
+            video_embeddings, text_embeddings, logit_scale = model_video(embeddings, toks, prenorm=False, postnorm=False)
```
Comment (iejMac) [inline src/training/train.py:92]: same here prenorm and postnorm True

File: .gitignore (hunk 14 above)
Comment (iejMac) [reply .gitignore:129]: ah this is just no \n at end of file

File: src/aggregation/aggregator_wrapper.py [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    def encode_text(self, x, postnorm=True):
+        with torch.no_grad():
+            text_embeddings = self.model_text(x).float()
+        return F.normalize(text_embeddings, dim=-1) if postnorm else text_embeddings
+
+    def encode_video(self, x, prenorm=True, postnorm=True):
+        x = F.normalize(x.float(), dim=-1) if prenorm else x
+        x = self.aggregator(x)
+        x = F.normalize(x.float(), dim=-1) if postnorm else x
+        return x, self.logit_scale.exp()
```
Comment (iejMac) [inline src/aggregation/aggregator_wrapper.py:49]: no need to return logit_scale in encode_video, just in forward

File: src/aggregation/aggregator_wrapper.py [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+    def forward(self, video_embeddings, toks, prenorm=True, postnorm=True):
+        text_embeddings = self.encode_text(toks, postnorm=postnorm)
+        video_embeddings, logit_scale = self.encode_video(video_embeddings, prenorm=prenorm, postnorm=postnorm)
+        return video_embeddings, text_embeddings, logit_scale
```
Comment (iejMac) [inline src/aggregation/aggregator_wrapper.py:54]: self.logit_scale.exp()

File: src/aggregation/factory.py [hunk 20]
```diff
@@ -54,4 +56,4 @@ def create_model(cfg_path, pretrained=''):
             logging.warning(f'Pretrained weights ({pretrained}) not found for model {model_name}.')
             raise RuntimeError(f'Pretrained weights ({pretrained}) not found for model {model_name}.')
 
-    return model, get_model_string(model_dict)
+    return VideoCLIP(model, clip_model), get_model_string(model_dict)
```
Comment (iejMac) [inline src/aggregation/factory.py:59]: newline at end

File: src/evaluation/retrieval.py [hunk 21]
```diff
@@ -31,8 +31,7 @@ def retrieval_evaluation(model_video, model_text, data, multicaption=False):
             embeddings = embeddings.to(device, non_blocking=True)
             toks = toks.to(device, non_blocking=True)
 
-            video_embeddings = model_video(embeddings)
-            text_embeddings = model_text(toks)
+            video_embeddings, text_embeddings, logit_scale = model_video(embeddings, toks, prenorm=True, postnorm=True)
```
Comment (iejMac) [inline src/evaluation/retrieval.py:34]: since we don't use this logit_scale you can just have that position be _

File: src/training/train.py [hunk 22]
```diff
@@ -145,4 +137,4 @@ def get_metrics(video_features, text_features, logit_scale):
         for k in [1, 5, 10]:
             metrics[f"{name}_R@{k}"] = np.mean(preds < k)
 
-    return metrics
+    return metrics
```
Comment (iejMac) [inline src/training/train.py:140]: newline

**Comment by iejMac** [comment]
well done Daniel, LGTM
